tests/ui/macros/macro-metavar-expr-concat/syntax-errors.rs

```diff
@@ -43,8 +37,69 @@ macro_rules! dollar_sign_without_referenced_ident {
     };
 }
 
+macro_rules! starting_number {
+    ($ident:ident) => {{
+        let ${concat(1, $ident)}: () = ();
+        //~^ ERROR `${concat(..)}` is not generating a valid identifier
+    }};
+}
+
+macro_rules! starting_valid_unicode {
+    ($ident:ident) => {{
+        let ${concat('Ý', $ident)}: () = ();
+    }};
+}
+
+macro_rules! starting_invalid_unicode {
+    ($ident:ident) => {{
+        let ${concat("\u{999999}", $ident)}: () = ();
+        //~^ ERROR invalid unicode character escape
```

Comment:
Where is this error coming from?
I don't see any additional checking in transcribe.rs that would result in this.

Also, will `concat!("\x41")` produce `a` or an invalid identifier `\x41`?

Reply:
> Where is this error coming from?

https://github.com/rust-lang/rust/blob/8a8ad3433e4168f54ac1364f85da694be9eeca7c/compiler/rustc_parse/src/lexer/unescape_error_reporting.rs#L36-L41

Looks like the parse eagerly halts when an invalid unicode literal is found. Changed to "\u{00BD}" (½).

> Also, will concat!("\x41") produce a or an invalid identifier \x41?

At the current time it is not possible to accept literal parameters but I expect to see a `${concat(..)} is not generating a valid identifier` error like in https://github.com/rust-lang/rust/pull/126841/files#diff-c527d06b3a89d2cf1181cf0b3122091572d2992346b42217dc290a18d2e2366cR75

cc https://github.com/rust-lang/rust/pull/126841#discussion_r1659970139

Reply:
Okay, the current behavior looks acceptable to me - either we have a literal that doesn't have `\` in it so its escaped and unescaped values are the same, or it has `\` and then it's not a valid identifier and we produce an error.